{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c02647e2-d884-464d-9663-439e203c0cba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app '__main__'\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on all addresses (0.0.0.0)\n",
      " * Running on http://127.0.0.1:5001\n",
      " * Running on http://192.168.1.30:5001\n",
      "Press CTRL+C to quit\n",
      "192.168.1.18 - - [19/Aug/2024 21:32:21] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:32:22] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:32:25] \"POST /delete/OWNER/4 HTTP/1.1\" 400 -\n",
      "[2024-08-19 21:33:42,333] ERROR in app: Exception on / [GET]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/pi/.local/lib/python3.11/site-packages/mysql/connector/connection_cext.py\", line 705, in cmd_query\n",
      "    self._cmysql.query(\n",
      "_mysql_connector.MySQLInterfaceError: Table 'test123.your_table_name' doesn't exist\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3/dist-packages/flask/app.py\", line 2525, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/flask/app.py\", line 1822, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/flask/app.py\", line 1820, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/lib/python3/dist-packages/flask/app.py\", line 1796, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/tmp/ipykernel_8983/1748417140.py\", line 21, in index\n",
      "    cursor.execute('SELECT * FROM your_table_name')  # Replace with your actual table name\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/pi/.local/lib/python3.11/site-packages/mysql/connector/cursor_cext.py\", line 357, in execute\n",
      "    result = self._connection.cmd_query(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/pi/.local/lib/python3.11/site-packages/mysql/connector/opentelemetry/context_propagation.py\", line 97, in wrapper\n",
      "    return method(cnx, *args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/pi/.local/lib/python3.11/site-packages/mysql/connector/connection_cext.py\", line 713, in cmd_query\n",
      "    raise get_mysql_exception(\n",
      "mysql.connector.errors.ProgrammingError: 1146 (42S02): Table 'test123.your_table_name' doesn't exist\n",
      "192.168.1.18 - - [19/Aug/2024 21:33:42] \"GET / HTTP/1.1\" 500 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:34:58] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:17] \"POST /add_row/OWNER HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:17] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:34] \"POST /add_row/CATEGORY HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:35] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:41] \"POST /modify/CATEGORY/6 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:42] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:46] \"POST /delete/CATEGORY/6 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:47] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:54] \"POST /modify/OWNER/5 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:35:54] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:06] \"POST /modify/OWNER/3 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:06] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:26] \"POST /add_row/DOC_STATE HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:27] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:38] \"POST /add_row/REVIEW_STATE HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:39] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:54] \"POST /modify/DOCUMENT/3 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:54] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:59] \"POST /modify/DOCUMENT/3 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:36:59] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:04] \"POST /modify/DOCUMENT/3 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:04] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:09] \"POST /modify/DOCUMENT/3 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:10] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:14] \"POST /delete/OWNER/5 HTTP/1.1\" 400 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:20] \"POST /modify/DOCUMENT/3 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:21] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:23] \"POST /delete/OWNER/5 HTTP/1.1\" 302 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:23] \"GET /configuration HTTP/1.1\" 200 -\n",
      "192.168.1.18 - - [19/Aug/2024 21:37:36] \"GET /document HTTP/1.1\" 404 -\n"
     ]
    }
   ],
   "source": [
    "from flask import Flask, render_template, request, redirect, url_for\n",
    "import mysql.connector\n",
    "from mysql.connector import Error\n",
    "\n",
    "\n",
    "app = Flask(__name__)\n",
    "\n",
    "def get_db_connection():\n",
    "    conn = mysql.connector.connect(\n",
    "        host=\"localhost\",\n",
    "        user=\"admin\",\n",
    "        password=\"helloLayka\",\n",
    "        database=\"test123\"\n",
    "    )\n",
    "    return conn\n",
    "\n",
    "\n",
    "\n",
    "@app.route('/configuration')\n",
    "def configuration():\n",
    "    conn = get_db_connection()\n",
    "    cursor = conn.cursor(dictionary=True)\n",
    "\n",
    "    # Fetching all rows for OWNER, CATEGORY, DOC_STATE, REVIEW_STATE\n",
    "    cursor.execute('SELECT * FROM OWNER')\n",
    "    owner_rows = cursor.fetchall()\n",
    "\n",
    "    cursor.execute('SELECT * FROM CATEGORY')\n",
    "    category_rows = cursor.fetchall()\n",
    "\n",
    "    cursor.execute('SELECT * FROM DOC_STATE')\n",
    "    doc_state_rows = cursor.fetchall()\n",
    "\n",
    "    cursor.execute('SELECT * FROM REVIEW_STATE')\n",
    "    review_state_rows = cursor.fetchall()\n",
    "\n",
    "    # Query to join DOCUMENT with OWNER and CATEGORY tables\n",
    "    cursor.execute('''\n",
    "        SELECT DOCUMENT.id, DOCUMENT.name, DOCUMENT.initials, OWNER.name AS owner_name, CATEGORY.name AS category_name, DOCUMENT.owner_id, DOCUMENT.category_id\n",
    "        FROM DOCUMENT\n",
    "        JOIN OWNER ON DOCUMENT.owner_id = OWNER.id\n",
    "        JOIN CATEGORY ON DOCUMENT.category_id = CATEGORY.id\n",
    "    ''')\n",
    "    document_rows = cursor.fetchall()\n",
    "\n",
    "    # Fetching lists for dropdowns in the DOCUMENT table\n",
    "    cursor.execute('SELECT id, name FROM OWNER')\n",
    "    owner_list = cursor.fetchall()\n",
    "\n",
    "    cursor.execute('SELECT id, name FROM CATEGORY')\n",
    "    category_list = cursor.fetchall()\n",
    "\n",
    "    cursor.close()\n",
    "    conn.close()\n",
    "\n",
    "    return render_template('configuration.html',\n",
    "                           owner_rows=owner_rows, \n",
    "                           category_rows=category_rows, \n",
    "                           doc_state_rows=doc_state_rows, \n",
    "                           review_state_rows=review_state_rows,\n",
    "                           document_rows=document_rows,\n",
    "                           owner_list=owner_list,\n",
    "                           category_list=category_list)\n",
    "\n",
    "@app.route('/modify/<table_name>/<int:row_id>', methods=['POST'])\n",
    "def modify_row(table_name, row_id):\n",
    "    conn = get_db_connection()\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # Build the update query dynamically based on the form fields\n",
    "    update_fields = ', '.join([f\"{key} = %s\" for key in request.form.keys()])\n",
    "    values = list(request.form.values())\n",
    "\n",
    "    cursor.execute(f\"UPDATE {table_name} SET {update_fields} WHERE id = %s\", values + [row_id])\n",
    "    conn.commit()\n",
    "    cursor.close()\n",
    "    conn.close()\n",
    "\n",
    "    return redirect(url_for('configuration'))\n",
    "\n",
    "\n",
    "\n",
    "@app.route('/delete/<table_name>/<int:row_id>', methods=['POST'])\n",
    "def delete_row(table_name, row_id):\n",
    "    conn = get_db_connection()\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    try:\n",
    "        cursor.execute(f'DELETE FROM {table_name} WHERE id = %s', (row_id,))\n",
    "        conn.commit()\n",
    "    except Error as e:\n",
    "        # Check if the error is due to a foreign key constraint\n",
    "        if e.errno == 1451:  # MySQL error code for foreign key constraint failure\n",
    "            cursor.close()\n",
    "            conn.close()\n",
    "            return f\"Cannot delete this item because it is still in use. The {table_name} entry with ID {row_id} could not be deleted because it is referenced by another table.\", 400\n",
    "        else:\n",
    "            raise e  # Re-raise if it's another kind of error\n",
    "    \n",
    "    cursor.close()\n",
    "    conn.close()\n",
    "    \n",
    "    return redirect(url_for('configuration'))\n",
    "\n",
    "@app.route('/add_row/<table_name>', methods=['POST'])\n",
    "def add_row(table_name):\n",
    "    conn = get_db_connection()\n",
    "    cursor = conn.cursor()\n",
    "\n",
    "    # Extract the form data, excluding the id field since it's usually auto-incremented\n",
    "    keys = ', '.join(request.form.keys())\n",
    "    placeholders = ', '.join(['%s'] * len(request.form))\n",
    "    values = list(request.form.values())\n",
    "\n",
    "    cursor.execute(f\"INSERT INTO {table_name} ({keys}) VALUES ({placeholders})\", values)\n",
    "    conn.commit()\n",
    "    cursor.close()\n",
    "    conn.close()\n",
    "\n",
    "    return redirect(url_for('configuration'))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run(host='0.0.0.0', port=5001)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04746edd-d745-4c63-be63-fded303f693e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
